sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/MutationGroupEncoder.java

```diff
@@ -0,0 +1,649 @@
+        }
+      } else {
+        Type.Code code = val.getType().getCode();
+        switch (code) {
+          case BOOL:
+            long v = val.getBool() ? 0 : 1;
+            if (part.isDesc()) {
+              orderedCode.writeSignedNumDecreasing(v);
+            } else {
+              orderedCode.writeSignedNumIncreasing(v);
+            }
+            break;
+          case INT64:
+            if (part.isDesc()) {
+              orderedCode.writeSignedNumDecreasing(val.getInt64());
+            } else {
+              orderedCode.writeSignedNumIncreasing(val.getInt64());
+            }
+            break;
+          case FLOAT64:
+            if (part.isDesc()) {
+              orderedCode.writeSignedNumDecreasing(Double.doubleToLongBits(val.getFloat64()));
+            } else {
+              orderedCode.writeSignedNumIncreasing(Double.doubleToLongBits(val.getFloat64()));
+            }
+            break;
+          case STRING:
+            if (part.isDesc()) {
+              orderedCode.writeBytesDecreasing(val.getString().getBytes());
+            } else {
+              orderedCode.writeBytes(val.getString().getBytes());
+            }
+            break;
+          case BYTES:
+            if (part.isDesc()) {
+              orderedCode.writeBytesDecreasing(val.getBytes().toByteArray());
+            } else {
+              orderedCode.writeBytes(val.getBytes().toByteArray());
+            }
+            break;
+          case TIMESTAMP: {
+            Timestamp value = val.getTimestamp();
+            if (part.isDesc()) {
+              orderedCode.writeNumDecreasing(value.getSeconds());
+              orderedCode.writeNumDecreasing(value.getNanos());
+            } else {
+              orderedCode.writeNumIncreasing(value.getSeconds());
+              orderedCode.writeNumIncreasing(value.getNanos());
+            }
+            break;
+          }
+          case DATE:
+            Date value = val.getDate();
+            if (part.isDesc()) {
+              orderedCode.writeSignedNumDecreasing(encodeDate(value));
+            } else {
+              orderedCode.writeSignedNumIncreasing(encodeDate(value));
+            }
+            break;
+          default:
+            throw new IllegalArgumentException("Unknown type " + val.getType());
+        }
+      }
+    }
+    return orderedCode.getEncodedBytes();
+  }
+
+  public byte[] encodeKey(String table, Key key) {
+    OrderedCode orderedCode = new OrderedCode();
+    List<SpannerSchema.KeyPart> parts = schema.getKeyParts(table);
+    Iterator<Object> it = key.getParts().iterator();
+    for (SpannerSchema.KeyPart part : parts) {
+      Object value = it.next();
+      if (value == null) {
+        if (part.isDesc()) {
+          orderedCode.writeInfinityDecreasing();
+        } else {
+          orderedCode.writeInfinity();
+        }
+      } else {
+        if (value instanceof Boolean) {
+          long v = (Boolean) value ? 0 : 1;
+          if (part.isDesc()) {
```

Comment:
Seems like this and similar code in the previous method should be extracted into a utility method that writes a Java Object to orderedCode in ascending or descending order.